### PR TITLE
ISSUE-122 Classifier Callback

### PR DESCRIPTION
**Pull request by thomkenn**
Early tentative formatting for final classifier callback, add an argument for the classifier type and allow it to process the image based on that.

### CODE REVIEW

File: pycoal/mineral.py [hunk 1]
```diff
@@ -62,14 +62,15 @@ def __init__(self, library_file_name, class_names=None, threshold=0.0, in_memory
          "-spectral library '%s', -class names '%s', -threshold '%s', -in_memory '%s'" 
             %(library_file_name, class_names, threshold, in_memory))
 
-    def classify_image(self, image_file_name, classified_file_name):
+    def classify_image(self, image_file_name, classified_file_name, classifier_type=None):
```
Comment (lewismc) [inline pycoal/mineral.py]: Change this parameter to default of ```SAM```

File: pycoal/mineral.py [hunk 2]
```diff
@@ -62,14 +62,15 @@ def __init__(self, library_file_name, class_names=None, threshold=0.0, in_memory
...
         """
-        Classify minerals in an AVIRIS image using spectral angle mapper
+        Classify minerals in an AVIRIS image using specified classifier_type, if none defaults to spectral angle mapper
         classification and save the results to a file.
 
         Args:
             image_file_name (str):      filename of the image to be classified
             classified_file_name (str): filename of the classified image
+            classifier_type (str):      Classifier type used
```
Comment (lewismc) [inline pycoal/mineral.py]: Augment this with the list of possible classifier options e.g. SAM, GML, KMeans, MDC (MahalanobisDistanceClassifier), etc. More classifiers can be seen at http://www.spectralpython.net/class_func_glossary.html#classification

File: pycoal/mineral.py [hunk 3]
```diff
@@ -91,44 +92,64 @@ def classify_image(self, image_file_name, classified_file_name):
         # TODO band resampler should do this
         resample = spectral.BandResampler([x/1000 for x in image.bands.centers],
                                           self.library.bands.centers)
-
+        if classifier_type == None or classifier_type == 'SAM':
```
Comment (lewismc) [inline pycoal/mineral.py]: Default value should always be SAM, remove reference to None

File: pycoal/mineral.py [hunk 4]
```diff
@@ -91,44 +92,64 @@ def classify_image(self, image_file_name, classified_file_name):
...
+
+                        # classify pixel if confidence above threshold
+                        if angles[0,0,index_of_max] > self.threshold:
+
+                            # index from one (after zero for no data)
+                            classified[x,y] = index_of_max + 1
+
+        elif classifier_type == 'GML':
+            # Gaussian Maximum Likelihood
+            # the following is the method layed out in the spectral python documentation, needs to be converted to work with our formatting
```
Comment (lewismc) [inline pycoal/mineral.py:130]: Don't just comment that it needs to be converted... actually go ahead and convert it.

File: pycoal/mineral.py [hunk 5]
```diff
@@ -91,44 +92,64 @@ def classify_image(self, image_file_name, classified_file_name):
...
+            # classify our training image and display the resulting classification map.
+            clmap = gmlc.classify_image(image)
+            v = imshow(classes=clmap)
+            # mask out all the pixels not associated with a training class
+            gtresults = clmap * (gt != 0)
+            v = imshow(classes=gtresults)
+            
+        elif classifier_type == 'KMeans':
+            # K-means generates clusters, this will read in pixels and figure out nearest cluster center
+            # the following is the method layed out in the spectral python documentation, needs to be converted to work with our formatting
```
Comment (lewismc) [inline pycoal/mineral.py]: Same applies here... go ahead and implement the classification don't just say you will... do it.

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @thomkenn thank you fro submitting pull request. Please consider my feedback. Thanks.

**Comment by lewismc** [comment]
Hi @thomkenn any update? Haven;t heard anything in almost a week! Are you stuck with something?

**Comment by lewismc** [comment]
@thomkenn PING this issue hasn't been touched in over a week. We agreed that the communication aspect of this work would improve... do you have a progress update please? Thanks

**Comment by thomkenn** [comment]
Ill admit, im having some issue with this section thats taking me a bit longer than i had anticipated. The different classifier callbacks are going to have different return types and require different arguments, and the part about 'training initialization' is proving kind of difficult, im just trying to figure out how to work that into the function and still maintain its place within the rest of the project. This is my top priority right now, but ill strive for better communication in this pivotal time. I can do a quick pull request updating the stuff you identified earlier, and once i have my plan for dealing with the earlier problems i can push those as well

**Comment by lewismc** [comment]
> Ill admit, im having some issue with this section thats taking me a bit longer than i had anticipated. 

When we discussed what your commitments were going to be as per the requirements document, I stated that, based on previous performance, these issues would take longer than both you and Bryce thought they would. 

> The different classifier callbacks are going to have different return types and require different arguments, and the part about 'training initialization' is proving kind of difficult, im just trying to figure out how to work that into the function and still maintain its place within the rest of the project. 

It would have been good if you had communicated this earlier to be honest. We are 8 days on from my initial feedback and it doesn't look like anything further has been achieved. Please let me know if things are "...proving kind of difficult" and we can discuss them here.
 
> This is my top priority right now, but ill strive for better communication in this pivotal time.

The only way that we are going to get the pull requests to an acceptable state is if you communicate with me. There is unfortunately no other way. We discussed this at the previous meeting with Kevin and it is imperative we keep the communication a priority.

> I can do a quick pull request updating the stuff you identified earlier, and once i have my plan for dealing with the earlier problems i can push those as well

Yes please do the quick pull request and I will review again then we can iterate.

I want to say the same thing to you as I did to Bryce, ALWAYS test your pull request before submitting. This is a base requirement for any source code contribution. If the proposed code edit breaks the unit tests then it is unacceptable in any Software project anywhere. The same is true here and contributions need to ensure both backwards compatibility as well as ensuring tests pass...

Please tag me once you've submitted your quick pull request. We will take it from there.

**Comment by lewismc** [comment]
PING @thomkenn

**Comment by lewismc** [comment]
@EvandroCT please take a look at this PR and see if your NN plans would fit in here. Thank you

**Comment by lewismc** [comment]
Kenny, what is the issue you are having here? We need to hash this out... 
I’m not understanding what the issue is here. You need to communicate what the issue is.

**Comment by lewismc** [comment]
Take a look at the comments... you've not addressed them. If this is not done by Thursday morning I will just do it myself.

**Comment by thomkenn** [comment]
The most recent commit should address the quick issues we discussed earlier, this was the 'quick fix' i mentioned, just taking care of the the sam being the default option as well as the removing of the 'none' item. The other issues will require a reworking of the funciton as we know it to include the 'training' that these classifier methods talk about and a way to pass different arguments for different classifiers

**Comment by lewismc** [comment]
@thomkenn 

> The most recent commit should address the quick issues we discussed earlier, this was the 'quick fix' i mentioned, just taking care of the the sam being the default option as well as the removing of the 'none' item.

This is not exactly quick if it took a week to edit around 4 lines of code!

> The other issues will require a reworking of the funciton as we know it to include the 'training' that these classifier methods talk about and a way to pass different arguments for different classifiers

Is there any reason that this has turned out to be so difficult? It is pretty straight forward work with most of it already done.... Additionally, your PR breaks the existing tests... you've not dealt with that either. 

Can you please again consider that this is meant to be 8 hours worth of work. It's taken you 2 full weeks to submit a pull request which covers about 5 or so actual lines of source code. The rest is formatting. I don't understand what the hold up is here and you are not explaining it to me either so I am not able to assess what the real issue is. Please try and help me out here.

**Comment by lewismc** [comment]
Closing this off.

**Pull request closed without merging**